[Style] Convert `keyword | primitive` properties to use strong style types

weinig opened the pull request:
#### 1cce65ce10388636344eab7eeb19f37b858f53e4
<pre>
[Style] Convert `keyword | primitive` properties to use strong style types
<a href="https://bugs.webkit.org/show_bug.cgi?id=296413">https://bugs.webkit.org/show_bug.cgi?id=296413</a>

Reviewed by Darin Adler.

Adds a new aggregate, `ValueOrKeyword`, and uses it to implement strong style
types for `-webkit-hyphenate-limit-after`, `-webkit-hyphenate-limit-before`,
`-webkit-hyphenate-limit-lines`, `-webkit-hyphenate-character`, `max-lines`,
`-webkit-line-grid` and `block-step-size`.

Also simplified the implementations of `ClipEdge`, `ScrollbarColor` and `Perspective`
by adopting `ValueOrKeyword`.

* Source/WTF/wtf/text/AtomString.h:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSPrimitiveValue.h:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSValue.h:
* Source/WebCore/css/values/CSSValueAggregates.h:
* Source/WebCore/layout/formattingContexts/inline/InlineContentBreaker.cpp:
* Source/WebCore/layout/formattingContexts/inline/InlineFormattingContext.cpp:
* Source/WebCore/layout/formattingContexts/inline/InlineLayoutState.h:
* Source/WebCore/layout/formattingContexts/inline/TextOnlySimpleLineBuilder.cpp:
* Source/WebCore/page/LocalFrameViewLayoutContext.cpp:
* Source/WebCore/rendering/LineClampUpdater.h:
* Source/WebCore/rendering/RenderBlock.cpp:
* Source/WebCore/rendering/RenderBlockFlow.cpp:
* Source/WebCore/rendering/RenderLayoutState.cpp:
* Source/WebCore/rendering/RenderText.cpp:
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
* Source/WebCore/rendering/style/StyleRareInheritedData.h:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleBuilderCustom.h:
* Source/WebCore/style/StyleExtractorConverter.h:
* Source/WebCore/style/StyleExtractorCustom.h:
* Source/WebCore/style/StyleExtractorSerializer.h:
* Source/WebCore/style/StyleInterpolationWrappers.h:
* Source/WebCore/style/values/StyleValueTypes+CSSValueConversion.h:
* Source/WebCore/style/values/line-grid/StyleWebKitLineGrid.h: Added.
* Source/WebCore/style/values/masking/StyleClip.cpp:
* Source/WebCore/style/values/masking/StyleClip.h:
* Source/WebCore/style/values/overflow/StyleMaximumLines.h: Added.
* Source/WebCore/style/values/primitives/StylePrimitiveNumeric.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+CSSValueConversion.h:
* Source/WebCore/style/values/rhythm/StyleBlockStepSize.cpp: Added.
* Source/WebCore/style/values/rhythm/StyleBlockStepSize.h: Added.
* Source/WebCore/style/values/scrollbars/StyleScrollbarColor.cpp:
* Source/WebCore/style/values/scrollbars/StyleScrollbarColor.h:
* Source/WebCore/style/values/text/StyleHyphenateCharacter.h: Added.
* Source/WebCore/style/values/text/StyleHyphenateLimitEdge.h: Added.
* Source/WebCore/style/values/text/StyleHyphenateLimitLines.h: Added.
* Source/WebCore/style/values/transforms/StylePerspective.cpp:
* Source/WebCore/style/values/transforms/StylePerspective.h:

Canonical link: <a href="https://commits.webkit.org/297892@main">https://commits.webkit.org/297892@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b8cf8f00ec486bcfb24a3f9aa4f798adf4b431d3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/113268 "23 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/33008 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/23481 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/119476 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/63935 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/115230 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/33656 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/41566 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/86206 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/41215 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/116215 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/26884 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/101889 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/66532 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/26148 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/20017 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/63230 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/105785 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/96258 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/20092 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/122693 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/111884 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/40346 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/30098 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/95057 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/40731 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/98104 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/94803 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24193 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/39947 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/17763 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/36482 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/40232 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/45731 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/136114 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/39873 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/36521 "Found 2 new JSC stress test failures: wasm.yaml/wasm/stress/jitless-wasm-to-js-should-follow-to-wasm-call-frame-convention-2.js.default, wasm.yaml/wasm/stress/jitless-wasm-to-js-should-follow-to-wasm-call-frame-convention.js.default (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/43206 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/41610 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/css/values/CSSValueAggregates.h:
```diff
@@ -490,6 +490,52 @@ template<typename T> struct CommaSeparatedRefCountedFixedVector {
+    constexpr ValueOrKeyword(Keyword)
+        : m_value { }
+    {
+    }
```
```suggestion
    constexpr ValueOrKeyword(Keyword)
    {
    }
```

darinadler commented on Source/WebCore/rendering/LineClampUpdater.h:
```diff
@@ -58,10 +58,9 @@ inline LineClampUpdater::LineClampUpdater(const RenderBlockFlow& blockContainer)
+    if (auto maximumLinesForBlockContainer = m_blockContainer->style().maxLines().tryValue()) {
         // New, top level line clamp.
-        layoutState->setLineClamp(RenderLayoutState::LineClamp { maximumLinesForBlockContainer, m_blockContainer->style().overflowContinue() == OverflowContinue::Discard });
+        layoutState->setLineClamp(RenderLayoutState::LineClamp { static_cast<size_t>(maximumLinesForBlockContainer->value), m_blockContainer->style().overflowContinue() == OverflowContinue::Discard });
```
What type are we casting from here? Is there way to do this without `static_cast`?

weinig replied:
It's an `int`, but one that is guaranteed to never be negative (since it is from an `Style::Integer<CSS::Nonnegative, int>`). We don want to use a `Style::Integer<CSS::Nonnegative, size_t>`, since having 2^64 lines is silly, and we want the negative bits for `Markable`.

weinig replied:
We can probably change RenderLayoutState::LineClamp to use an `int` to avoid the cast, but I'll leave that later for now.

darinadler commented on Source/WebCore/rendering/style/RenderStyle.cpp:
```diff
@@ -2456,16 +2456,20 @@ CounterDirectiveMap& RenderStyle::accessCounterDirectives()
-
-    // FIXME: This should depend on locale.
     static MainThreadNeverDestroyed<const AtomString> hyphenMinusString(span(hyphenMinus));
     static MainThreadNeverDestroyed<const AtomString> hyphenString(span(hyphen));
```
Maybe we should move these inside the lambda down where we put the `FIXME` comment

darinadler commented on Source/WebCore/rendering/style/RenderStyle.cpp:
```diff
@@ -2456,16 +2456,20 @@ CounterDirectiveMap& RenderStyle::accessCounterDirectives()
+        },
+        [&](const AtomString& string) -> const AtomString& {
+            return string;
+        }
```
```suggestion
        [](const AtomString& string) -> const AtomString& {
            return string;
        }
```

darinadler commented on Source/WebCore/style/values/StyleValueTypes+CSSValueConversion.h:
```diff
@@ -24,12 +24,37 @@
+   {
+        if (!value.isString()) {
+            state.setCurrentPropertyInvalidAtComputedValueTime();
+            return emptyString();
```
Is empty better than null here? Sometimes we choose null in such cases because it’s slightly more efficient? But in other cases we use empty since null is such a peculiar value.

weinig replied:
It is unobservable, but since we use null internally to mean another thing, empty seems “safer”. But it’s a coin toss.

darinadler replied:
Empty “safer” is enough better than null that I say yes, it is better.

darinadler commented on Source/WebCore/style/values/StyleValueTypes+CSSValueConversion.h:
```diff
@@ -24,12 +24,37 @@
+   {
+        if (!value.isString()) {
+            state.setCurrentPropertyInvalidAtComputedValueTime();
+            return emptyAtom();
```
Is empty better than null here? Sometimes we choose null in such cases because it’s slightly more efficient? But in other cases we use empty since null is such a peculiar value.

darinadler commented on Source/WebCore/style/values/transforms/StylePerspective.h:
```diff
@@ -31,44 +31,22 @@ namespace Style {
+        if (auto value = tryValue())
+            return std::max(1.0f, value->value);
+        return 1.0f;
+    }
```
I’ve seen cases where we use `value_or` for this type of thing. Maybe less efficient.
```suggestion
    float usedPerspective() const { return std::max(1.0f, tryValue().value_or(1.0f)->value); }
```

weinig committed:
[Style] Convert `keyword | primitive` properties to use strong style types
https://bugs.webkit.org/show_bug.cgi?id=296413

Reviewed by Darin Adler.

Adds a new aggregate, `ValueOrKeyword`, and uses it to implement strong style
types for `-webkit-hyphenate-limit-after`, `-webkit-hyphenate-limit-before`,
`-webkit-hyphenate-limit-lines`, `-webkit-hyphenate-character`, `max-lines`,
`-webkit-line-grid` and `block-step-size`.

Also simplified the implementations of `ClipEdge`, `ScrollbarColor` and `Perspective`
by adopting `ValueOrKeyword`.

* Source/WTF/wtf/text/AtomString.h:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSPrimitiveValue.h:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSValue.h:
* Source/WebCore/css/values/CSSValueAggregates.h:
* Source/WebCore/layout/formattingContexts/inline/InlineContentBreaker.cpp:
* Source/WebCore/layout/formattingContexts/inline/InlineFormattingContext.cpp:
* Source/WebCore/layout/formattingContexts/inline/InlineLayoutState.h:
* Source/WebCore/layout/formattingContexts/inline/TextOnlySimpleLineBuilder.cpp:
* Source/WebCore/page/LocalFrameViewLayoutContext.cpp:
* Source/WebCore/rendering/LineClampUpdater.h:
* Source/WebCore/rendering/RenderBlock.cpp:
* Source/WebCore/rendering/RenderBlockFlow.cpp:
* Source/WebCore/rendering/RenderLayoutState.cpp:
* Source/WebCore/rendering/RenderText.cpp:
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
* Source/WebCore/rendering/style/StyleRareInheritedData.h:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleBuilderCustom.h:
* Source/WebCore/style/StyleExtractorConverter.h:
* Source/WebCore/style/StyleExtractorCustom.h:
* Source/WebCore/style/StyleExtractorSerializer.h:
* Source/WebCore/style/StyleInterpolationWrappers.h:
* Source/WebCore/style/values/StyleValueTypes+CSSValueConversion.h:
* Source/WebCore/style/values/line-grid/StyleWebKitLineGrid.h: Added.
* Source/WebCore/style/values/masking/StyleClip.cpp:
* Source/WebCore/style/values/masking/StyleClip.h:
* Source/WebCore/style/values/overflow/StyleMaximumLines.h: Added.
* Source/WebCore/style/values/primitives/StylePrimitiveNumeric.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+CSSValueConversion.h:
* Source/WebCore/style/values/rhythm/StyleBlockStepSize.cpp: Added.
* Source/WebCore/style/values/rhythm/StyleBlockStepSize.h: Added.
* Source/WebCore/style/values/scrollbars/StyleScrollbarColor.cpp:
* Source/WebCore/style/values/scrollbars/StyleScrollbarColor.h:
* Source/WebCore/style/values/text/StyleHyphenateCharacter.h: Added.
* Source/WebCore/style/values/text/StyleHyphenateLimitEdge.h: Added.
* Source/WebCore/style/values/text/StyleHyphenateLimitLines.h: Added.
* Source/WebCore/style/values/transforms/StylePerspective.cpp:
* Source/WebCore/style/values/transforms/StylePerspective.h:

Canonical link: https://commits.webkit.org/297892@main
